MetaMask/metamask-mobile
# fix: Yarn 1.22.22

## **Description**

Prevent engineers using different Yarn versions with different features
causing our CI dedupe step to break
- [x] Bump package.json Yarn engine from ^1.22.0 to 1.22.22 
- [x] Bump BitRise Yarn version from 1.22.19 to 1.22.22
- [x] Lock project Yarn version to 1.22.22
- [x] Add Yarn config file comment RE @lavamoat/allow-scripts setup
- [x] Run `yarn deduplicate` and commit
- [x] Revert Bump BitRise Yarn version from 1.22.19 to 1.22.22 ([thread](https://github.com/MetaMask/metamask-mobile/pull/9143#pullrequestreview-1980917241))
- [x] Bump BitRise Yarn version from 1.22.19 to 1.22.22
  - [x] Fix https://github.com/yarnpkg/yarn/issues/9050
  - [x] Switch to <s>NPM or</s> Corepack

Nb: our metamask-extension setup
- https://github.com/MetaMask/metamask-extension/blob/develop/package.json#L616
- https://github.com/MetaMask/metamask-extension/blob/develop/package.json#L681
- https://github.com/MetaMask/metamask-extension/blob/develop/.yarn/releases/yarn-4.0.2.cjs
- https://github.com/MetaMask/metamask-extension/blob/develop/.yarnrc.yml#L142

https://classic.yarnpkg.com/lang/en/docs/package-json/#toc-engines
> The engines specify versions of clients that must be used with your package. This checks against `process.versions` as well as the current version of yarn.

https://classic.yarnpkg.com/lang/en/docs/yarnrc/#toc-yarn-path
> Instructs yarn to defer to another Yarn binary for execution. Useful if you want to bundle Yarn into your repository and have everyone use the same version for consistency. This was introduced in Yarn 1.0, so all developers must have Yarn >= 1.0 installed.

https://yarnpkg.com/getting-started/qa

> `.yarn/plugins` and `.yarn/releases` contain the Yarn releases used in the current repository (as defined by [yarn ](https://yarnpkg.com/cli/set/version)[set version](https://yarnpkg.com/cli/set/version)). You will want to keep them versioned (this prevents potential issues if, say, two engineers use different Yarn versions with different features).

https://yarnpkg.com/cli/set/version

> Lock the Yarn version used by the project.

https://nodejs.org/dist/latest-v15.x/docs/api/all.html (introduced in Node v16 only)
https://nodejs.org/dist/latest-v16.x/docs/api/all.html#all_packages_packagemanager
from testing our repo (on Node v18), this appears to have no effect enforcing the version

from testing our repo, we could revert https://github.com/MetaMask/metamask-mobile/pull/9143/commits/c7d67e605e76f21ad8dda8af3da45355d8ea2056 to keep our structure simpler
since the package.json engine field is enough to enforce the version
but these 2 files are generated by running `yarn set version 1.22.22`

Yarn v1.22 update
- PR: https://github.com/yarnpkg/yarn/pull/9023
- Commit: https://github.com/yarnpkg/yarn/commit/88d5e44e3b64e54c82872cd8da07af5a068b1279
- Picked into branch: https://github.com/yarnpkg/yarn/tree/1.22-stable

## **Related issues**

Fixes: CI dedupe step consistently failing

## **Manual testing steps**

- ensure `yarn --version` is below 1.22.22
- or e.g. `yarn set version 1.22.21`
- run `yarn`
- Yarn errors correctly ^1.22.22+ is expected

Nb: after doing this too many times, you can use a VPN as a workaround

<img width="1806" alt="Screenshot 2024-04-04 at 6 40 35 PM" src="https://github.com/MetaMask/metamask-mobile/assets/1881059/41cf015c-2aed-4c63-8f24-041a5cf98ef3">

## **Screenshots/Recordings**

<img width="987" alt="Screenshot 2024-04-04 at 8 32 00 PM" src="https://github.com/MetaMask/metamask-mobile/assets/1881059/901ca091-02db-4a69-8137-0ec1b6b3c07d">

## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
